File: 229-majority-element-ii/majority-element-ii.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        int n = nums.size();
        int c1 = 0, e1;
        int c2 = 0, e2;

        for(int i=0;i<nums.size();i++)
        {
            if(c1==0 && e2!=nums[i])
            {
                c1 = 1;
                e1 = nums[i];
            }
            else if(c2==0 && e1!=nums[i])
            {
                c2 = 1;
                e2 = nums[i];
            }
            else if(e1 == nums[i])
            {
                c1++;
            }
            else if(e2 == nums[i])
            {
                c2++;
            }
            else
            {
                c1--;
                c2--;
            }
        }

        c1 = 0;
        c2 = 0;
        for(int i=0;i<n;i++)
        {
            if(nums[i]==e1)c1++;
            if(nums[i]==e2)c2++;
        }

        vector<int>ans;

        if(c1>=n/3+1)
        ans.push_back(e1);
        if(c2>=n/3+1)
        ans.push_back(e2);

        return ans;
        
    }
};

File: Difficulty: Medium/Majority Element II/majority-element-ii.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        sc.nextLine(); // Consume the newline character

        while (t-- > 0) {
            String s = sc.nextLine();
            String[] parts = s.split(" ");
            int[] nums = new int[parts.length];
            for (int i = 0; i < parts.length; i++) {
                nums[i] = Integer.parseInt(parts[i]);
            }
            Solution ob = new Solution();
            List<Integer> ans = ob.findMajority(nums);

            if (ans.size() == 0) {
                System.out.println("[]");
            } else {
                for (int i : ans) {
                    System.out.print(i + " ");
                }
                System.out.println();
            }
        }
        sc.close();
    }
}

// } Driver Code Ends

class Solution {
    public List<Integer> findMajority(int[] nums) {
        // Your code goes here.
        int n=nums.length;
        int one_third=n/3;
        HashMap<Integer,Integer> map=new HashMap<>();
        List<Integer> res=new ArrayList<>();
        
        for(int a:nums){
            map.put(a,map.getOrDefault(a,0)+1);
        }
        
        for (Integer key : map.keySet()) {
            if(map.get(key)>one_third)res.add(key);
        }
        
        return res;
    }
}

